### DISCUSSION TITLE
AppleTalk: Printing works, but no file server

### DISCUSSION THREAD

**Discussion by marcuskuenzel1971** [akuker/RASCSI#1112] [General]
i activated the piscsi_bridge option for the network configuration in the atalk.conf file
cups and papd is installed and running

I can now select my network printer in the chooser as laserwriter 8 an printing works.

But i can't select a file server in the chooser.

the services tell me the following:

- afpd.service - Netatalk AFP fileserver for Macintosh clients
     Loaded: loaded (/lib/systemd/system/afpd.service; enabled; vendor preset: enabled)
     Active: active (running) since Mon 2023-02-27 18:19:53 GMT; 24h ago
    Process: 479 ExecStart=/usr/local/sbin/afpd -c 20 (code=exited, status=0/SUCCESS)
      Tasks: 2 (limit: 407)
        CPU: 66ms
     CGroup: /system.slice/afpd.service
             └─484 /usr/local/sbin/afpd -c 20

Feb 27 18:19:53 piscsi afpd[484]: dsi_tcp: hostname 'piscsi' resolves to loopback address
Feb 27 18:19:53 piscsi afpd[484]: dsi_tcp (Chooser will not select afp/tcp) Check to make sure piscsi is in /etc/hosts and the correct domain is in /etc/resolv.conf: Invalid argument
Feb 27 18:19:53 piscsi afpd[484]: AFP/TCP started, advertising 0.0.0.0:548 (2.230201)
Feb 27 18:20:00 piscsi afpd[484]: uam: "DHX2" available
Feb 27 18:20:00 piscsi afpd[484]: uam: "Cleartxt Passwrd" available
Feb 27 18:20:00 piscsi afpd[484]: uam: "No User Authent" available
Feb 27 18:20:00 piscsi afpd[484]: Registering server 'piscsi' with Bonjour
Feb 27 18:20:00 piscsi afpd[484]: Successfully started avahi loop.
Feb 27 18:20:00 piscsi afpd[484]: Registering CNID module [last]
Feb 27 18:20:00 piscsi afpd[484]: Registering CNID module [dbd]

- atalkd.service - Netatalk AppleTalk daemon
     Loaded: loaded (/lib/systemd/system/atalkd.service; enabled; vendor preset: enabled)
     Active: active (running) since Tue 2023-02-28 18:32:06 GMT; 32min ago
    Process: 969 ExecStart=/usr/local/sbin/atalkd (code=exited, status=0/SUCCESS)
      Tasks: 1 (limit: 407)
        CPU: 57ms
     CGroup: /system.slice/atalkd.service
             └─970 /usr/local/sbin/atalkd

Feb 28 18:32:05 piscsi atalkd[970]: config for no router
Feb 28 18:32:06 piscsi atalkd[970]: ready 0/0/0
Feb 28 18:32:06 piscsi systemd[1]: Started Netatalk AppleTalk daemon.
Feb 28 18:49:10 piscsi atalkd[970]: nbp_packet: malformed packet
Feb 28 18:49:10 piscsi atalkd[970]: nbp_packet: malformed packet
Feb 28 18:49:11 piscsi atalkd[970]: nbp_packet: malformed packet
Feb 28 18:49:12 piscsi atalkd[970]: nbp_packet: malformed packet
Feb 28 18:49:13 piscsi atalkd[970]: nbp_packet: malformed packet
Feb 28 18:49:14 piscsi atalkd[970]: nbp_packet: malformed packet
Feb 28 18:49:15 piscsi atalkd[970]: nbp_packet: malformed packet

- /etc/hosts has this:
127.0.0.1       localhost
::1             localhost ip6-localhost ip6-loopback
ff02::1         ip6-allnodes
ff02::2         ip6-allrouters
127.0.1.1               piscsi

Is there anything wrong with the hosts-configuration ?

**Comment by rdmark** [comment]
@marcuskuenzel1971 The hosts message is a red herring I think. Service discovery in the Chooser is entirely AppleTalk, which doesn't care about TCP settings. I'm more concerned about the `nbp_packet: malformed packet` messages in the atalkd log. I've never seen that before. What output do you get from `nbplkup`?

**Comment by marcuskuenzel1971** [comment]
it works now, after i added a parameter in the conf-file.

But after the boot of the pi the services sometimes do not start correctly so i have the to restart manually

**Comment by rdmark** [comment]
@marcuskuenzel1971 May I ask which parameter you added to which conf file, just for my (and other user's) reference?

As to the issue of services not starting up, this may be improved by this PR that I filed today: https://github.com/Netatalk/netatalk/pull/233

**Comment by marcuskuenzel1971** [comment]
I added -setuplog "default log_maxdebug /var/log/afpd.log" to the - -transall -uamlist uams_guest.so,uams_clrtxt.so,uams_dhx2.so -nosavepassword -noicon option in /etc/netatalk/afpd.conf

**Comment by rdmark** [comment]
@marcuskuenzel1971 I see, thanks for sharing. My guess is that the act of restarting afpd was what fixed the problem. In order to register as an AppleTalk service, atalkd must have started up and accepting registrations before you start afpd. Sometimes systemd doesn't do a good job with starting them in the right order. The solution is when atalkd is running, to stop afpd, wait a few seconds, then start afpd again.

**Comment by marcuskuenzel1971** [comment]
> As to the issue of services not starting up, this may be improved by this PR that I filed today: https://github.com/Netatalk/netatalk/pull/233

where can i find distrib/initscripts/afpd.service.tmpl and edit it ?

**Comment by rdmark** [comment]
I responded over in the upstream PR

**Comment by marcuskuenzel1971** [comment]
there is also the following problem:

i added the "piscsi_bridge" in the /etc/netatalk/atalkd.conf file and the service starts with this configuration and adds independently some parameters behind the option. AppleTalk works with this configuration.

but after the next boot the service independently replaces the "piscsi_bridge" to the "wlan0" interface in the /etc/netatalk/atalkd.conf file, AppleTalk doesnt work. I have to manually edit the /atalkd.conf again with the piscsi_bridge interface

**Reply by rdmark** [reply]
So atalkd does try to be "smart" and automatically configure the interfaces that it sees.

Are you using the wlan0 interface for anything? Do you have wired ethernet too? One option is to completely disable wifi on your RPi.

**Comment by marcuskuenzel1971** [comment]
i edit now the service files, but for atalk.service after boot it shows the following status:

● atalkd.service - Netatalk AppleTalk daemon
     Loaded: loaded (/lib/systemd/system/atalkd.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Wed 2023-03-01 21:09:51 GMT; 5min ago
    Process: 491 ExecStart=/usr/local/sbin/atalkd (code=exited, status=1/FAILURE)
        CPU: 10ms

Mar 01 21:09:51 piscsi systemd[1]: atalkd.service: Scheduled restart job, restart counter is at 5.
Mar 01 21:09:51 piscsi systemd[1]: Stopped Netatalk AppleTalk daemon.
Mar 01 21:09:51 piscsi systemd[1]: atalkd.service: Start request repeated too quickly.
Mar 01 21:09:51 piscsi systemd[1]: atalkd.service: Failed with result 'exit-code'.
Mar 01 21:09:51 piscsi systemd[1]: Failed to start Netatalk AppleTalk daemon.

**Reply by rdmark** [reply]
Can you please share the contents of /lib/systemd/system/atalkd.service ?

**Comment by marcuskuenzel1971** [comment]
Finally I managed it to work.

The problem is that the atalk, afpd and papd services don't start properly as long as all necessary network conections are ready after boot. Also the piscsi_bridge network must run. Otherwise the atalk-service selects the wlan interface automaticaly, but with this interface the mac can't see the AppleTalk-servers. 

So i added some options in the service file for atalk:

Requires=network-online.service
After=network-online.service

BindsTo=sys-devices-virtual-net-lan0.device
After=sys-devices-virtual-net-lan0.device

and "network-online.service" also in the service files for afpd and papd.
